src/I18n/PackageLocator.php

```diff
@@ -0,0 +1,124 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @copyright     Copyright (c) 2017 Aura for PHP
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.2.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\I18n;
+
+use Cake\I18n\Exception\I18nException;
+
+/**
+ * A ServiceLocator implementation for loading and retaining package objects.
+ *
+ * @internal
+ */
+class PackageLocator
+{
+    /**
+     * A registry of packages.
+     *
+     * Unlike many other registries, this one is two layers deep. The first
+     * key is a package name, the second key is a locale code, and the value
+     * is a callable that returns a Package object for that name and locale.
+     *
+     * @var array
+     */
+    protected $registry = [];
+
+    /**
+     * Tracks whether or not a registry entry has been converted from a
+     * callable to a Package object.
+     *
+     * @var array
+     */
+    protected $converted = [];
+
+    /**
+     * Constructor.
+     *
+     * @param array $registry A registry of packages.
+     * @see $registry
+     */
+    public function __construct(array $registry = [])
+    {
+        foreach ($registry as $name => $locales) {
+            foreach ($locales as $locale => $spec) {
+                $this->set($name, $locale, $spec);
+            }
+        }
+    }
+
+    /**
+     * Sets a Package loader.
+     *
+     * @param string $name The package name.
+     * @param string $locale The locale for the package.
+     * @param callable|\Cake\I18n\Package $spec A callable that returns a package or Package instance.
+     * @return void
+     */
+    public function set(string $name, string $locale, $spec): void
```

Comment:
I guess this is required for compatibility for `set` and `setPackage` are confusingly similar.

Reply:
Initially I had thought of adding a separate method for adding `Package` instance so that the argument could be typehinted for both methods but later just updated `set()` to accept both types to keep code DRY and forgot to remove the `setPackage()` method.

What does everyone prefer? Single method on two separate ones? I can update accordingly.